### ISSUE TITLE
[!] Failed to "make install"! Please have a look, Thanks!

### ISSUE THREAD

**Issue by MuSamDu** [qterminal/qterminal#168]
I'm trying to install your qterminal. I installed all the dependences, and successfully compiled it. But when i try to "make install" it, errors occured.
Outputs after "make install" begin to seem strange from here:
>
[ 54%] Building CXX object CMakeFiles/qterminal.dir/src/main.cpp.o
[ 57%] Building CXX object CMakeFiles/qterminal.dir/src/mainwindow.cpp.o
[ 59%] Building CXX object CMakeFiles/qterminal.dir/src/tabwidget.cpp.o
[ 61%] Building CXX object CMakeFiles/qterminal.dir/src/termwidget.cpp.o
/home/msd/Downloads/qterminal/src/termwidget.cpp:107:2: warning: #warning TODO/FIXME: disable the action when there is only one terminal [-Wcpp]
 #warning TODO/FIXME: disable the action when there is only one terminal
  ^
[ 64%] Building CXX object CMakeFiles/qterminal.dir/src/termwidgetholder.cpp.o
[ 66%] Building CXX object CMakeFiles/qterminal.dir/src/properties.cpp.o
[ 69%] Building CXX object CMakeFiles/qterminal.dir/src/propertiesdialog.cpp.o
[ 71%] Building CXX object CMakeFiles/qterminal.dir/src/bookmarkswidget.cpp.o
[ 73%] Building CXX object CMakeFiles/qterminal.dir/src/fontdialog.cpp.o
[ 76%] Building CXX object CMakeFiles/qterminal.dir/src/third-party/qxtglobalshortcut.cpp.o
[ 78%] Building CXX object CMakeFiles/qterminal.dir/src/third-party/qxtglobalshortcut_x11.cpp.o
/home/msd/Downloads/qterminal/src/third-party/qxtglobalshortcut_x11.cpp:37:45: fatal error: qpa/qplatformnativeinterface.h: No such file or directory
 #   include <qpa/qplatformnativeinterface.h>
                                             ^
compilation terminated.
CMakeFiles/qterminal.dir/build.make:407: recipe for target 'CMakeFiles/qterminal.dir/src/third-party/qxtglobalshortcut_x11.cpp.o' failed
make[2]: *** [CMakeFiles/qterminal.dir/src/third-party/qxtglobalshortcut_x11.cpp.o] Error 1
CMakeFiles/Makefile2:122: recipe for target 'CMakeFiles/qterminal.dir/all' failed
make[1]: *** [CMakeFiles/qterminal.dir/all] Error 2
Makefile:136: recipe for target 'all' failed
make: *** [all] Error 2

Please tell me what the hell is going on and try to solve it, Thanks a lot!

**Comment by antis81** [comment]
That comes absolutely natural! The reason: libQxt tries to pull in private Qt headers - very bad idea! If such a thing is required, better copy the headers. I investigated further and found, this was some interims state, as QX11Info got lost in the early days of Qt5. We should forget about that -> it's not required anymore.

What we need here is to bring the module `QtX11Extras` and the header `#include <QX11Info>` back into play. I just changed that locally, getting my stuff run.

A more generic question: What is the "global keyboard shortcuts" concept even used for? Isn't there a better solution, than copying libQxt files? They're just existing in `QTerminal` for this reason. We actually should remove that dependency.

**Comment by jleclanche** [comment]
Global shortcuts is used for the qterminal dropdown shortcut iirc

**Comment by antis81** [comment]
@jleclanche Ok ... that makes it a bit more clear, thanks.

If I understand correctly, the `F12` global shortcut is meant to open a qterminal dropdown (which doesn't work for me however). LXQt manages it's global shortcuts via `lxqt-globalkeys`. Shouldn't that be rather put there?

Next questions:
* Shouldn't global shortcuts be managed solely by the desktop environment?
* What would happen, if two app's register for the same shortcut?

Could you please have a look at `m_dropShortcut` located in `mainwindow.h`? This is the only place, where `QxtGlobalShortcut` is used. IMHO we should actually get rid of that somehow.

**Comment by pvanek** [comment]
at first - you need to start terminal in dropdown mode (-d or by launcher)

I don't use dropdown mode at all but it seems like quite lots of people like this feature, so we cannot simple remove it.

- handling global shorcuts by the environment would be super-great. But how we can achieve it? We have KDE, Gnome, OS X, LxQt, plain window managers... I really wish there can be something what can do this work for us but I don't know about anything yet.

- if there are 2 conflicting keyboard shortcuts registered, only the 1st one is used, 2nd one is rejected resulting in some warning message. The app is running but global shortcut is not working for the 2nd app

**Comment by antis81** [comment]
>at first - you need to start terminal in dropdown mode (-d or by launcher)

This doesn't even work on my system (Lubuntu 15.04 without `lxqt-panel`).

>I don't use dropdown mode at all but it seems like quite lots of people like this feature, so we cannot simple remove it.

I'm not talking about removing the `dropdown terminal` feature itself. I don't use it either, but I saw it in action on youtube.

As so often, I'm looking at several puzzled pieces here, which I can hardly express. But let me proceed ...

First, I stumbled upon this very issue by myself - I also couldn't compile. So I analysed the current implementation of the global shortcut and found it's a crappy mess - I mean it. This needs to change. Let me prioritize steps.

## Step 1
The most important thing is to get `QTerminal` compiling for others, being able to proceed as quickly as possible. Thus, we **must** get rid of that private Qt header! This is a heavy bug introduced from the third-party `libQxt`. Don't EVER refer to private headers! Qt put a big warning there, that someone decided to ignore! This actually disqualifies code for productive use. Let's make it short: I have fixed that locally and will setup another PR soon ...

## Step 2
We need to discuss about how to do it right. AFAIK there's no established standard concept for global shortcut registration. Anyway, we can make the best out of it and implement DE dependent global key registration - with focus on the main target platform (LXQt). Maybe we can take another application - e.g. `lxqt-runner` as inspiration. Changing the "start" shortcut (default: `Alt+F2`) is reflected in `lxqt-globalkeys` (the global shortcut manager).

Screenshot of `lxqt-runner` and `lxqt-config-globalshortcuts`:

![lxqt-runner-hotkey](https://cloud.githubusercontent.com/assets/440517/9292956/4c985b48-4415-11e5-968c-dd671f8b5a09.png)

If looking closely, you will notice a shortcut to fire up `QTerminal` (default: `Ctrl+Alt+T`), but we can't see it in QTerminal settings. Further we cannot see the `F12` shortcut - which is what we want in the first place.

We should actually be able to implement that for several (major) desktop environments (e.g. LXQt/LXDE, KDE, Gnome, XFCE, ??? and probably OSX), respecting their mechanisms. An XDG standard would help greatly!

As long as that is not there ... How about simply adding a shortcut to the DE's global hotkey manager (something like: `F12 => run "qterminal -d"`).

Sorry for the long reply - and also for emotions probably included. I think it makes sense to setup another issue (start a new task) for that.

**Comment by pvanek** [comment]
ad step1 - there are many Qt5 apps using private headers too. So I'd like to re-qualify the phrase *Thus, we must get rid of that private Qt header!* to *it would be good to get rid of that private Qt header*.
Currently I don't see any way how to do it in really multiplatform way, except to provide eg. D-BUS interface for qterminal and leave user to configure it on his own. DBUS APi can provide method "toggle dropdown window"... But in this case I bet there will be many bugreports with "I cannot open drop down terminal, it's broken, fix it asap turbo++" because of lack of configuration for user's favourite DE/WM.

Also I don't want to spent many man-hours of working in not so important stuff (for me) so it can be taken as open for public development. What I can implement really quickly is a compilation check for the required private header(s) in the cmake configuration phase with extensive error message if it's not found. There can be an cmake option to disable features depending on that private headers too.

Private headers are available in repositories for all distros I use so I don't see it as a big issue.

ad step2 - this solution is LxQt only and it requires running lxqt-globalkeys daemon. So it's not a real way to go here because it won't work in KDE, etc. out of the box.
Don't take me wrong, please. I'm LxQt user, but I also use qterminal on OSX (which is one of real reason why I started to maintain qterminal)

So conclusion is - there is no standard way how to handle global shortcuts. And (at least for me) it looks like waste of time to implement something new while current code works. Sort of.

**Comment by pvanek** [comment]
see commit 7d740b00fa6fb274f2d1622a71bd2e4d20551b31
it results in standard error message if private headers are not found:

```
CMake Error at /usr/lib64/cmake/Qt5Gui/Qt5GuiConfig.cmake:27 (message):
  The imported target "Qt5::Gui" references the file

     "/usr/include/qt5/QtGui/5.4.2"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/lib64/cmake/Qt5Gui/Qt5GuiConfig.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/lib64/cmake/Qt5Gui/Qt5GuiConfig.cmake:72 (_qt5_Gui_check_file_exists)
  CMakeLists.txt:20 (find_package)


-- Configuring incomplete, errors occurred!
```

**Comment by antis81** [comment]
>Private headers are available in repositories for all distros I use so I don't see it as a big issue.

Private implementations (aka pimpl) are private for a reason. The term "private" means: This code is **not** part of (Qt) API. Translation: Don't (never ever) touch it!

Actually the libQxt project made a mistake by ignoring that rule and simply let go! Meanwhile Qt fixed the root of the problem (version 5.1) - libQxt still doesn't ... and ... :boom:.

If a programmer feels in urgent need of using third-party pimpl stuff in a project, the ONLY way is to copy headers plus implementation into own code - or better rewrite it.

Enforcing programmers to install private Qt headers via CMake rule makes it even worse. Please remove that.

Like I said: I solved the problem (in our libQxt version) ... PR is about to land soon.

**Comment by antis81** [comment]
Ths can be closed. @MuSamDu Solved! :)
